activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -73,11 +73,29 @@ def key_conversion_required?
           end
 
           def convert_key(key)
```

Comment:
After discussing we thought about moving the hashing functions to the column object so we can keep the knowledge of now to deal with case insensitive values in the column itself. Can you change this PR to delegate this logic to the column object?

Reply:
@rafaelfranca I'm not sure what you mean by the hashing functions. Are you thinking of something like this?

```diff
diff --git a/activerecord/lib/active_record/associations/preloader/association.rb b/activerecord/lib/active_record/associations/preloader/association.rb
index b61a5f0146..2f1dfdc089 100644
--- a/activerecord/lib/active_record/associations/preloader/association.rb
+++ b/activerecord/lib/active_record/associations/preloader/association.rb
@@ -83,19 +83,15 @@ def convert_key(key)
           end
 
           def case_sensitive_comparison?
-            association_key_case_sensitive? || owner_key_case_sensitive?
+            association_key_column.case_sensitive? || owner_key_column.case_sensitive?
           end
 
-          def association_key_case_sensitive?
-            case_sensitive_column?(klass.columns_hash[association_key_name.to_s])
+          def association_key_column
+            klass.columns_hash[association_key_name.to_s]
           end
 
-          def owner_key_case_sensitive?
-            case_sensitive_column?(model.columns_hash[owner_key_name.to_s])
-          end
-
-          def case_sensitive_column?(column)
-            column.respond_to?(:case_sensitive?) ? column.case_sensitive? : true
+          def owner_key_column
+            model.columns_hash[owner_key_name.to_s]
           end
 
           def association_key_type
diff --git a/activerecord/lib/active_record/connection_adapters/column.rb b/activerecord/lib/active_record/connection_adapters/column.rb
index 5d81de9fe1..bfb76f96d0 100644
--- a/activerecord/lib/active_record/connection_adapters/column.rb
+++ b/activerecord/lib/active_record/connection_adapters/column.rb
@@ -74,6 +74,10 @@ def hash
         attributes_for_hash.hash
       end
 
+      def case_sensitive?
+        true
+      end
+
       protected
 
         def attributes_for_hash

```

Reply:
@rafaelfranca are you able to shed some more light on the intended direction so this can be brought over the line. I am getting bit by this now by joins on a `citext` column with varying cases. Our case is joining email addresses on accounts to email addresses on project invitations, which are entered by humans and are allowed to vary by cases and we want to preserve that variance for display purposes.

I'm not aware of a workaround so if this patch is close to how you envision, I'd consider picking it up. If it's vary different, though, I probably can't as I am no longer familiar enough with modern AR internals.

Reply:
The hashing function is the `convert_key` method. So our idea is to move `convert_key` implementation to the column. So instead of doing `owners_by_key[convert_key(record[association_key_name])]` we can do `owners_by_key[record.get_cache_key_for(association_key_name)]`